### PR TITLE
Support for passing 'task' in -Daemon option

### PR DESCRIPTION
**Pull request by NimishaS**
## Description
Currently `-Daemon` option accepts only none and service. It should accept `task` too.

---

This checklist is used to make sure that common guidelines for a pull request are followed. You can find a more complete discussion of PowerShell cmdlet best practices [here](https://msdn.microsoft.com/en-us/library/dd878270(v=vs.85).aspx).

- [x] **I have read the [contribution guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md).**
- [ ] **If changes were made to any cmdlet, the XML help was regenerated using the [platyPSHelp module](https://github.com/Azure/azure-powershell/blob/dev/documentation/platyPSHelp-documentation.md).**
- [x] **If any large changes are made to a service, they are reflected in the respective [change log](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#updating-the-change-log).**

### [General Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#general-guidelines)
- [x] Title of the pull request is clear and informative.
- [x] There are a small number of commits, each of which have an informative message. This means that previously merged commits do not appear in the history of the PR. For more information on cleaning up the commits in your PR, [see this page](https://github.com/Azure/azure-powershell/blob/dev/documentation/cleaning-up-commits.md).
- [x] The pull request does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/dev/documentation/breaking-changes.md) (unless a major version change occurs in the assembly and module).

### [Testing Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#testing-guidelines)
- [x] Pull request includes test coverage for the included changes.
- [x] PowerShell scripts used in tests should do any necessary setup as part of the test or suite setup, and should not use hard-coded values for locations or existing resources.

### [Cmdlet Signature Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-signature-guidelines)
- [x] New cmdlets that make changes or have side effects should implement `ShouldProcess` and have `SupportShouldProcess=true` specified in the cmdlet attribute. You can find more information on `ShouldProcess` [here](https://gist.github.com/markcowl/338e16fe5c8bbf195aff9f8af0db585d#what-is-the-change).
- [ ] Cmdlet specifies `OutputType` attribute if any output is produced - if the cmdlet produces no output, it should implement a `PassThru` parameter.

### [Cmdlet Parameter Guidelines](https://github.com/Azure/azure-powershell/blob/dev/CONTRIBUTING.md#cmdlet-parameter-guidelines)
- [x] Parameter types should not expose types from the management library - complex parameter types should be defined in the module.
- [x] Complex parameter types are discouraged - a parameter type should be simple types as often as possible. If complex types are used, they should be shallow and easily creatable from a constructor or another cmdlet.
- [ ] Cmdlet parameter sets should be mutually exclusive - each parameter set must have at least one mandatory parameter not in other parameter sets.

### CODE REVIEW

**Comment by azurecla** [comment]
Hi __@NimishaS__, I'm your friendly neighborhood Azure Pull Request Bot (You can call me AZPRBOT). Thanks for your contribution!
    <span>You've already signed the contribution license agreement. Thanks!</span>
        <span>We will now validate the agreement and then real humans will evaluate your PR.</span>

TTYL, AZPRBOT;

**Comment by cormacpayne** [comment]
@NimishaS please fill out the checklist at the top of the PR. Also, there is no test coverage for this added value for the `Daemon` parameter

File: src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/Chef/SetAzureVMChefExtension.cs [hunk 1]
```diff
@@ -107,9 +107,10 @@ public class SetAzureVMChefExtensionCommand : VirtualMachineChefExtensionCmdletB
         [Parameter(
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "Configures the chef-client service for unattended execution. The node platform should be Windows." +
-                          "Options: 'none' or 'service'." +
+                          "Allowed options: 'none', 'service' and 'task'" +
                           "none - Currently prevents the chef-client service from being configured as a service." +
-                          "service - Configures the chef-client to run automatically in the background as a service.")]
+                          "service - Configures the chef-client to run automatically in the background as a service." +
+                           "task - Configures the chef-client to run automatically in the background as a secheduled task.")]
         [ValidateNotNullOrEmpty]
```
Comment (shahabhijeet) [inline src/ServiceManagement/Compute/Commands.ServiceManagement/IaaS/Extensions/Chef/SetAzureVMChefExtension.cs]: @NimishaS can we add a ValidateSet attribute for this. That will drastically reduce a lot of code that is written to verify if it has the right values, plus would avoid non-supported values.
https://msdn.microsoft.com/en-us/library/system.management.automation.validatesetattribute(v=vs.85).aspx

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs [hunk 2]
```diff
@@ -119,9 +119,10 @@ public string TypeHandlerVersion
         [Parameter(
             ValueFromPipelineByPropertyName = true,
             HelpMessage = "Configures the chef-client service for unattended execution. The node platform should be Windows." +
-                          "Options: 'none' or 'service'." +
```
Comment (shahabhijeet) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs]: @NimishaS can we get rid of lot of code that is written again for null checking in ValidateParameters function, almost all the parameters are marked as
[ValidateNotNullOrEmpty]
consider using parameter sets which is really good to manage your long list of parameters.
Also consider using default value attributes if applicable for any of the parameters. That will also reduce code that is written to set defaults.

**Comment by NimishaS** [comment]
@shahabhijeet , I have added `[ValidateSet]` for the newly added parameter. I went through the chef extension code for both `ServiceManagement` and `ResourceManager` to identify if `[DefaultValue]` and `[ValidateSet]` can be used somewhere. I found that at most of the places the values which are set as default are fetched dynamically, hence `[DefaultValue]` can't be used.
I haven't added the test coverage for `ResourceManager` as I am getting `The current SynchronizationContext may not be used as a TaskScheduler` in record mode. We had faced this issue previously also: https://github.com/Azure/azure-powershell/issues/3341
The suggestions given on the issue didn't help us.

**Comment by NimishaS** [comment]
@shahabhijeet , can you please share the Travis failure errors?

**Comment by NimishaS** [comment]
@shahabhijeet , @cormacpayne , the testcases ran successfully on my local machine in `playback` mode. But the travis is still failing. Please share the details of the failure.

File: src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs (hunk 2 above)
Comment (NimishaS) [inline src/ResourceManager/Compute/Commands.Compute/Extension/Chef/SetAzureVMChefExtension.cs]: @shahabhijeet , please refer https://github.com/Azure/azure-powershell/pull/3500#issuecomment-280322934

**Comment by NimishaS** [comment]
Regarding Resource Manager test case, I have added a comment on the previous issue: https://github.com/Azure/azure-powershell/issues/3341#issuecomment-281324405

**Comment by shahabhijeet** [comment]
@NimishaS for the synchronization issue, look for test framework 1.5.4 that I will be releasing shortly.

**Comment by shahabhijeet** [comment]
@NimishaS static analysis failed. Take a look at the .csv files and let us know if you need any help resolving static analysis issues.

**Comment by NimishaS** [comment]
@shahabhijeet , I don't have access to see the details of Travis failures. So I can't see any .csv file. Please share all the failure messages.
It would be good if we can get access to check the Travis details. It would reduce some delays in completion of the task because right now we have to wait for someone's response if there is a failure.

**Comment by NimishaS** [comment]
@shahabhijeet , please share the travis failure details. I will fix the issues.
Regarding ARM testcases, if the release of  test framework 1.5.4 is going to take some time, can this PR be merged without ARM testcases since we need this functionality to be available to the users.
If the release is going to take one or two days, then we can wait.

**Comment by NimishaS** [comment]
@shahabhijeet , any updates here?

**Comment by markcowl** [comment]
@NimishaS 
You need to add exceptions for the new ValidateSet options you have added - I assume that scripts that use values outside these will not work.

Here are the needed entries, you can just add them in: https://github.com/Azure/azure-powershell/blob/dev/tools/StaticAnalysis/Exceptions/BreakingChangeIssues.csv 

```
"D:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Compute\Microsoft.Azure.Commands.Compute.dll","Microsoft.Azure.Commands.Compute.Extension.Chef.SetAzureVMChefExtensionCommand","Set-AzureRmVMChefExtension","0","2050","A validate set has been added for parameter 'Daemon' for cmdlet 'Set-AzureRmVMChefExtension'.","Remove the validate set from parameter 'Daemon'."
"D:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Compute\Microsoft.Azure.Commands.Compute.dll","Microsoft.Azure.Commands.Compute.Extension.Chef.SetAzureVMChefExtensionCommand","Set-AzureRmVMChefExtension","0","2090","The ValidateNotNullOrEmpty attribute has been added to parameter 'Linux' for cmdlet 'Set-AzureRmVMChefExtension'.","Remove the ValidateNotNullOrEmpty attribute from parameter 'Linux'."
"D:\workspace\powershell\src\Package\Debug\ResourceManager\AzureResourceManager\AzureRM.Compute\Microsoft.Azure.Commands.Compute.dll","Microsoft.Azure.Commands.Compute.Extension.Chef.SetAzureVMChefExtensionCommand","Set-AzureRmVMChefExtension","0","2090","The ValidateNotNullOrEmpty attribute has been added to parameter 'Windows' for cmdlet 'Set-AzureRmVMChefExtension'.","Remove the ValidateNotNullOrEmpty attribute from parameter 'Windows'."
```

**Comment by NimishaS** [comment]
Thanks @markcowl. Travis is passing now.

**Pull request merged**